compiler/aarch64/codegen/BinaryEvaluator.cpp

```diff
@@ -1153,16 +1299,153 @@ logicBinaryEvaluator(TR::Node *node, TR::InstOpCode::Mnemonic regOp, TR::InstOpC
+            {
+            /*
+             * The result is always 0 in this case.
+             */
+            TR::Register *reg = cg->allocateRegister();
+            generateTrg1ImmInstruction(cg, TR::InstOpCode::movzx, andNode, reg, 0);
+            andNode->setRegister(reg);
+            cg->recursivelyDecReferenceCount(shiftNode);
+            cg->decReferenceCount(maskNode);
+            return reg;
+            }
+
+         uint64_t shiftedMask = (maskValue >> shiftValue);
+         if (((shiftedMask & 1) == 1) && contiguousBits(shiftedMask))
+            {
+            /*
+             * shiftedMask has consecutive 1s from the least significant bits.
+             * So, this shift and mask operation is copying consecutive bits starting from the least significant bits of the source register
+             * to bit position shiftValue of the destination register.
+             */
+            uint32_t width = populationCount(shiftedMask);
+            uint32_t imms = (width - 1);
+            uint32_t immr = (is64bit ? 64 : 32) - shiftValue;
+
+            TR::Register *reg = cg->allocateRegister();
+            TR::Register *sreg = cg->evaluate(shiftNode->getFirstChild());
+
+            generateTrg1Src1ImmInstruction(cg, (is64bit ? TR::InstOpCode::ubfmx : TR::InstOpCode::ubfmw), andNode, reg, sreg, (immr << 6) | imms);
+            andNode->setRegister(reg);
+            cg->recursivelyDecReferenceCount(shiftNode);
+            cg->decReferenceCount(maskNode);
+            return reg;
+            }
+         }
+      else /* right shift*/
+         {
+         /*
+          * If the lsb of maskValue is set and the msb is not set and the maskValue has consecutive 1s, then this operation is copying 
+          * consecutive 1s starting from the bit position shiftValue of the source register to the least significant bits of the destination register.
+          * We consider arithmetic shift only.
+          */
+         if ((!shiftNode->getOpCode().isShiftLogical()) && ((maskValue & 1) == 1) && (((maskValue >> (is64bit ? 63 : 31)) & 1) == 0) && contiguousBits(maskValue))
+            {
+            uint32_t width = populationCount(maskValue);
+            uint32_t shiftRemainderWidth = (is64bit ? 64 : 32) - shiftValue;
+
+            if (width > shiftRemainderWidth)
+               {
+               if (shiftNode->getOpCode().isShiftLogical())
+                  {
+                  /* If it is a logical shift, we can ignore the mask beyond the msb. */
+                  width = shiftRemainderWidth;
+                  }
+               else
+                  {
+                  return NULL;
+                  }
+               }
+            bool isLogicalShiftRight = (width == shiftRemainderWidth);  /* In this case, it works as a logical shift right. */
+
+            uint32_t imms = (shiftValue + width - 1);
+            uint32_t immr = shiftValue;
+            TR::Register *reg = cg->allocateRegister();
```

Comment:
The same comment for the reference count of `sourceNode`.

Reply:
force-pushed the updated commit, too.